File: backend/app/schemas/command_meeting.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

MeetingStatus = Literal["sap_dien_ra", "dang_dien_ra", "da_ket_thuc", "da_huy"]
AttendanceStatus = Literal["co_mat", "vang_mat", "chua_diem_danh"]


class CommandMeetingCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    start_time: datetime
    end_time: Optional[datetime] = None
    location: Optional[str] = Field(default=None, max_length=255)
    meeting_link: Optional[str] = Field(default=None, max_length=500)
    agenda: Optional[str] = None
    attendee_user_ids: list[int] = Field(default_factory=list)


class CommandMeetingUpdate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    start_time: datetime
    end_time: Optional[datetime] = None
    location: Optional[str] = Field(default=None, max_length=255)
    meeting_link: Optional[str] = Field(default=None, max_length=500)
    agenda: Optional[str] = None
    status: MeetingStatus = "sap_dien_ra"


class MinutesRequest(BaseModel):
    minutes: str = Field(..., min_length=1)
    mark_finished: bool = False


class InviteRequest(BaseModel):
    user_ids: list[int] = Field(..., min_length=1)


class AttendanceUpdate(BaseModel):
    attendance: Optional[AttendanceStatus] = None
    absence_reason: Optional[str] = Field(default=None, max_length=500)
    contribution_note: Optional[str] = None


class AttendeeOut(BaseModel):
    id: int
    meeting_id: int
    user_id: int
    full_name: str
    unit_id: Optional[int]
    unit_name: Optional[str]
    attendance: AttendanceStatus
    absence_reason: Optional[str]
    contribution_note: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class CommandMeetingOut(BaseModel):
    id: int
    title: str
    start_time: datetime
    end_time: Optional[datetime]
    location: Optional[str]
    meeting_link: Optional[str]
    agenda: Optional[str]
    minutes: Optional[str]
    attachment_url: Optional[str]
    attachment_name: Optional[str]
    status: MeetingStatus
    classification: str
    created_by_id: int
    created_by_full_name: str
    created_at: datetime
    attendee_count: int
    present_count: int
    my_attendance: Optional[AttendanceStatus]

    model_config = ConfigDict(from_attributes=True)


class CommandMeetingDetailOut(CommandMeetingOut):
    attendees: list[AttendeeOut]

File: frontend/src/types/commandMeeting.ts
export type MeetingStatus = 'sap_dien_ra' | 'dang_dien_ra' | 'da_ket_thuc' | 'da_huy'
export type AttendanceStatus = 'co_mat' | 'vang_mat' | 'chua_diem_danh'

export const MEETING_STATUS_LABELS: Record<MeetingStatus, string> = {
  sap_dien_ra: 'Sắp diễn ra',
  dang_dien_ra: 'Đang diễn ra',
  da_ket_thuc: 'Đã kết thúc',
  da_huy: 'Đã huỷ',
}

export const ATTENDANCE_LABELS: Record<AttendanceStatus, string> = {
  co_mat: 'Có mặt',
  vang_mat: 'Vắng mặt',
  chua_diem_danh: 'Chưa điểm danh',
}

export interface Attendee {
  id: number
  meeting_id: number
  user_id: number
  full_name: string
  unit_id: number | null
  unit_name: string | null
  attendance: AttendanceStatus
  absence_reason: string | null
  contribution_note: string | null
}

export interface CommandMeeting {
  id: number
  title: string
  start_time: string
  end_time: string | null
  location: string | null
  meeting_link: string | null
  agenda: string | null
  minutes: string | null
  attachment_url: string | null
  attachment_name: string | null
  status: MeetingStatus
  classification: string
  created_by_id: number
  created_by_full_name: string
  created_at: string
  attendee_count: number
  present_count: number
  my_attendance: AttendanceStatus | null
}

export interface CommandMeetingDetail extends CommandMeeting {
  attendees: Attendee[]
}

export interface CommandMeetingCreate {
  title: string
  start_time: string
  end_time?: string | null
  location?: string | null
  meeting_link?: string | null
  agenda?: string | null
  attendee_user_ids: number[]
}

export interface CommandMeetingUpdate {
  title: string
  start_time: string
  end_time?: string | null
  location?: string | null
  meeting_link?: string | null
  agenda?: string | null
  status: MeetingStatus
}

export interface AttendanceUpdate {
  attendance?: AttendanceStatus
  absence_reason?: string | null
  contribution_note?: string | null
}
